In [1]:
import findspark
findspark.init()

In [2]:
from pyspark import SparkContext
sc = SparkContext("local", "first app")

## Module

Jaccard 계산용 math 라이브러리 <br>
CFG 클래스 : Threshold 등 계산시 필요한 상수값을 편하게 변경하기 위함

In [4]:
import math

In [5]:
class CFG : 
    sigma = 0.5

## Friends List

raw_fr : 날 것 그대로의 친구 데이터 <br>
fr_split : 데이터를 (key, value) pair로 변경 <br>
> lambda 함수를 이용한 단순 변경

friends_list : fr_split은 (자신, 친구) pair이므로 groupByKey를 통해 list화하여 친구 리스트 작성 <br>

In [3]:
raw_fr = sc.textFile("facebook_combined.txt")

In [6]:
raw_fr.take(10)

['0 1', '0 2', '0 3', '0 4', '0 5', '0 6', '0 7', '0 8', '0 9', '0 10']

In [7]:
fr_split = raw_fr.map(lambda x : x.split(' '))

In [8]:
fr_split = fr_test.map(lambda x : (int(x[0]), int(x[1])))

In [9]:
fr_split.take(10)

[(0, 1),
 (0, 2),
 (0, 3),
 (0, 4),
 (0, 5),
 (0, 6),
 (0, 7),
 (0, 8),
 (0, 9),
 (0, 10)]

In [10]:
friends_list = fr_split.groupByKey().mapValues(list)

In [11]:
friends_list.take(10)

[(0,
  [1,
   2,
   3,
   4,
   5,
   6,
   7,
   8,
   9,
   10,
   11,
   12,
   13,
   14,
   15,
   16,
   17,
   18,
   19,
   20,
   21,
   22,
   23,
   24,
   25,
   26,
   27,
   28,
   29,
   30,
   31,
   32,
   33,
   34,
   35,
   36,
   37,
   38,
   39,
   40,
   41,
   42,
   43,
   44,
   45,
   46,
   47,
   48,
   49,
   50,
   51,
   52,
   53,
   54,
   55,
   56,
   57,
   58,
   59,
   60,
   61,
   62,
   63,
   64,
   65,
   66,
   67,
   68,
   69,
   70,
   71,
   72,
   73,
   74,
   75,
   76,
   77,
   78,
   79,
   80,
   81,
   82,
   83,
   84,
   85,
   86,
   87,
   88,
   89,
   90,
   91,
   92,
   93,
   94,
   95,
   96,
   97,
   98,
   99,
   100,
   101,
   102,
   103,
   104,
   105,
   106,
   107,
   108,
   109,
   110,
   111,
   112,
   113,
   114,
   115,
   116,
   117,
   118,
   119,
   120,
   121,
   122,
   123,
   124,
   125,
   126,
   127,
   128,
   129,
   130,
   131,
   132,
   133,
   134,
   135,
   136,
   137,
   138,

In [12]:
len(friends_list.collect())

3663

## freq vocab

facebook_combined 데이터는 단방향 친구 데이터임 <br>
따라서 freq vocab을 만들기 위해서는 전체 빈도를 계산하기위해 쌍방향 친구 데이터로 재구성이 필요함 <br>

way2_map_func : (0, 1)이면 (0, 1), (1, 0)을 반환하여 쌍방향 친구관계를 구축하는 map 함수 <br>
way2_friends : map함수가 (자신, 친구) pair를 만들도록 설계되었으므로 groupByKey를 통해 모아줌 <br>
number_of_friends : 전체 친구 리스트에서 등장 빈도를 파악 <br>
vocab_list : 등장 빈도가 낮은 순으로 정렬하여 최종 결과인 freq_vocab_list를 제작 <br>

In [13]:
def way2_map_func(input_data) : 
    rnum, vec = input_data
    
    tmp_list = []
    for element in vec : 
        tmp_list.append((rnum, element))
        tmp_list.append((element, rnum))
    
    return tmp_list

In [14]:
way2_friends_map = friends_list.flatMap(way2_map_func)

In [15]:
way2_friends = way2_friends_map.groupByKey().mapValues(list)

In [16]:
way2_friends.take(10)

[(0,
  [1,
   2,
   3,
   4,
   5,
   6,
   7,
   8,
   9,
   10,
   11,
   12,
   13,
   14,
   15,
   16,
   17,
   18,
   19,
   20,
   21,
   22,
   23,
   24,
   25,
   26,
   27,
   28,
   29,
   30,
   31,
   32,
   33,
   34,
   35,
   36,
   37,
   38,
   39,
   40,
   41,
   42,
   43,
   44,
   45,
   46,
   47,
   48,
   49,
   50,
   51,
   52,
   53,
   54,
   55,
   56,
   57,
   58,
   59,
   60,
   61,
   62,
   63,
   64,
   65,
   66,
   67,
   68,
   69,
   70,
   71,
   72,
   73,
   74,
   75,
   76,
   77,
   78,
   79,
   80,
   81,
   82,
   83,
   84,
   85,
   86,
   87,
   88,
   89,
   90,
   91,
   92,
   93,
   94,
   95,
   96,
   97,
   98,
   99,
   100,
   101,
   102,
   103,
   104,
   105,
   106,
   107,
   108,
   109,
   110,
   111,
   112,
   113,
   114,
   115,
   116,
   117,
   118,
   119,
   120,
   121,
   122,
   123,
   124,
   125,
   126,
   127,
   128,
   129,
   130,
   131,
   132,
   133,
   134,
   135,
   136,
   137,
   138,

In [17]:
number_of_friends = way2_friends.map(lambda x : (x[0], len(x[1])))

In [18]:
number_of_friends.take(10)

[(0, 347),
 (1, 17),
 (2, 10),
 (3, 17),
 (4, 10),
 (5, 13),
 (6, 6),
 (7, 20),
 (8, 8),
 (9, 57)]

In [19]:
len(number_of_friends.collect())

4039

In [20]:
vocab_list = number_of_friends.sortBy(lambda x : x[1])

In [21]:
vocab_list.collect()

[(11, 1),
 (12, 1),
 (15, 1),
 (18, 1),
 (37, 1),
 (43, 1),
 (74, 1),
 (114, 1),
 (209, 1),
 (210, 1),
 (215, 1),
 (287, 1),
 (292, 1),
 (335, 1),
 (911, 1),
 (918, 1),
 (1096, 1),
 (1119, 1),
 (1145, 1),
 (1206, 1),
 (1386, 1),
 (1466, 1),
 (1560, 1),
 (1581, 1),
 (1834, 1),
 (358, 1),
 (447, 1),
 (550, 1),
 (585, 1),
 (602, 1),
 (607, 1),
 (608, 1),
 (613, 1),
 (624, 1),
 (638, 1),
 (668, 1),
 (674, 1),
 (692, 1),
 (801, 1),
 (875, 1),
 (883, 1),
 (891, 1),
 (892, 1),
 (2842, 1),
 (3031, 1),
 (3071, 1),
 (3183, 1),
 (3230, 1),
 (2079, 1),
 (2195, 1),
 (2269, 1),
 (2457, 1),
 (2470, 1),
 (2569, 1),
 (2596, 1),
 (3451, 1),
 (3453, 1),
 (3570, 1),
 (3650, 1),
 (3709, 1),
 (3729, 1),
 (3748, 1),
 (3798, 1),
 (3820, 1),
 (3853, 1),
 (3856, 1),
 (3935, 1),
 (3974, 1),
 (3984, 1),
 (4008, 1),
 (4010, 1),
 (4015, 1),
 (4022, 1),
 (4024, 1),
 (4035, 1),
 (33, 2),
 (35, 2),
 (42, 2),
 (47, 2),
 (52, 2),
 (70, 2),
 (90, 2),
 (138, 2),
 (145, 2),
 (153, 2),
 (154, 2),
 (160, 2),
 (183, 2),
 (205

In [22]:
freq_vocab_list = vocab_list.map(lambda x : x[0]).collect()

In [23]:
# 정렬용 broadcast
broad_vocab = sc.broadcast(freq_vocab_list)

In [24]:
broad_vocab.value

[11,
 12,
 15,
 18,
 37,
 43,
 74,
 114,
 209,
 210,
 215,
 287,
 292,
 335,
 911,
 918,
 1096,
 1119,
 1145,
 1206,
 1386,
 1466,
 1560,
 1581,
 1834,
 358,
 447,
 550,
 585,
 602,
 607,
 608,
 613,
 624,
 638,
 668,
 674,
 692,
 801,
 875,
 883,
 891,
 892,
 2842,
 3031,
 3071,
 3183,
 3230,
 2079,
 2195,
 2269,
 2457,
 2470,
 2569,
 2596,
 3451,
 3453,
 3570,
 3650,
 3709,
 3729,
 3748,
 3798,
 3820,
 3853,
 3856,
 3935,
 3974,
 3984,
 4008,
 4010,
 4015,
 4022,
 4024,
 4035,
 33,
 35,
 42,
 47,
 52,
 70,
 90,
 138,
 145,
 153,
 154,
 160,
 183,
 205,
 216,
 233,
 234,
 241,
 244,
 255,
 256,
 267,
 279,
 282,
 286,
 305,
 316,
 2814,
 1065,
 1176,
 1224,
 1233,
 1262,
 1270,
 1276,
 1326,
 1394,
 1395,
 1430,
 1558,
 1690,
 1755,
 1854,
 2704,
 2740,
 377,
 379,
 401,
 443,
 468,
 581,
 642,
 648,
 1967,
 699,
 744,
 749,
 750,
 775,
 841,
 859,
 860,
 2788,
 2857,
 3006,
 3125,
 3134,
 3268,
 3282,
 3407,
 3408,
 2382,
 2541,
 3589,
 3688,
 3704,
 3732,
 3746,
 3807,
 3808,
 3846,

## Sorting and Cut friends list

Inverted List를 만들 때 빈도가 낮은 순으로 정렬하여 일정한 규칙에 따른 개수로 잘라 만들어야함 <br>

cut_sort_by_list : 리스트를 정렬하여 |𝑥| − |𝜎 ⋅ |𝑥|| + 1인 prefix_size만큼 잘라 반환함
> if prefix_size == len(tmp_list) : break 부분은 필요 이상의 연산을 줄이기위한 체크

In [25]:
# sorting by another array values
# Z = [x for _,x in sorted(zip(Y,X))]

In [26]:
test = [78, 152, 181, 195, 218, 273, 275, 306, 328] # 4
len(test)

9

In [27]:
# list 값을 기준으로 sorting하는 방법이 필요
# 자동으로 prefix size |x| - |sigma - |x|| + 1로 나오도록 변경
def cut_sort_by_list(val_list, vocab_list) : 
    tmp_list = []
    len_val_list = len(val_list)
    prefix_size = math.floor(len_val_list - CFG.sigma * len_val_list + 1)
    for val in vocab_list : 
        if val in val_list : 
            tmp_list.append(val)
            if prefix_size == len(tmp_list) : 
                break
    return tmp_list

In [28]:
cut_sort_by_list(test, freq_vocab_list)

[152, 78, 195, 218, 273]

In [29]:
cut_sort_friends_list = friends_list.map(lambda x : (x[0], cut_sort_by_list(x[1], broad_vocab.value)))

In [30]:
broad_vocab.unpersist()

In [31]:
cut_sort_friends_list.take(10)

[(0,
  [11,
   12,
   15,
   18,
   37,
   43,
   74,
   114,
   209,
   210,
   215,
   287,
   292,
   335,
   33,
   35,
   42,
   47,
   52,
   70,
   90,
   138,
   145,
   153,
   154,
   160,
   183,
   205,
   216,
   233,
   234,
   241,
   244,
   255,
   256,
   267,
   279,
   282,
   286,
   305,
   316,
   61,
   71,
   76,
   81,
   97,
   112,
   155,
   157,
   164,
   179,
   182,
   191,
   207,
   228,
   240,
   247,
   253,
   293,
   294,
   301,
   321,
   336,
   49,
   120,
   124,
   125,
   166,
   174,
   190,
   201,
   202,
   206,
   220,
   262,
   270,
   288,
   289,
   307,
   327,
   27,
   34,
   46,
   110,
   152,
   192,
   193,
   235,
   245,
   250,
   264,
   283,
   6,
   32,
   44,
   63,
   77,
   86,
   95,
   102,
   117,
   147,
   163,
   219,
   229,
   269,
   340,
   51,
   64,
   83,
   126,
   129,
   131,
   151,
   167,
   189,
   208,
   237,
   263,
   296,
   300,
   311,
   317,
   338,
   347,
   8,
   54,
   60,
   89,
  

## Inverted map

빈도가 낮은 순으로 정렬되고 길이가 |𝑥| − |𝜎 ⋅ |𝑥|| + 1인 friends_list를 사용하여 inverted list 제작 <br>
invert_map_func : (자신, [친구 리스트])를 받아 각 친구에 대해 (친구, 자신) pair를 반환 <br>
> 따라서 groupByKey를 통해 흩어진 key-value pair를 모을 필요가 있음

inverted_friends_list에서 (1, [친구 리스트])가 없는 이유는 1번이 |𝑥| − |𝜎 ⋅ |𝑥|| + 1에서 잘렸기 때문

In [32]:
def invert_map_func(input_data) : 
    rnum, vec = input_data
    
    tmp_list = []
    for element in vec : 
        tmp_list.append((element, rnum))
    
    return tmp_list

In [33]:
inverted_friends_list_map = cut_sort_friends_list.flatMap(invert_map_func)

In [34]:
inverted_friends_list = inverted_friends_list_map.groupByKey().mapValues(list).sortByKey()

In [35]:
inverted_friends_list.collect()

[(2, [0]),
 (4, [0]),
 (6, [0]),
 (8, [0]),
 (10, [0]),
 (11, [0]),
 (12, [0]),
 (15, [0]),
 (16, [0]),
 (18, [0]),
 (19, [17]),
 (20, [2, 14]),
 (22, [0, 7]),
 (27, [0]),
 (28, [14]),
 (29, [16]),
 (30, [9]),
 (31, [21, 25]),
 (32, [0]),
 (33, [0]),
 (34, [0]),
 (35, [0]),
 (36, [0]),
 (37, [0]),
 (38, [0, 7]),
 (39, [21, 25]),
 (42, [0, 33]),
 (43, [0]),
 (44, [0, 20, 41]),
 (46, [0]),
 (47, [0]),
 (48, [1, 30]),
 (49, [0]),
 (50, [0]),
 (51, [0, 23, 25, 31]),
 (52, [0]),
 (54, [0, 1, 27, 48, 53]),
 (55, [21, 26]),
 (57, [24, 48]),
 (59, [13, 56]),
 (60, [0, 56]),
 (61, [0, 23]),
 (62, [26, 56]),
 (63, [0, 56]),
 (64, [0]),
 (65, [7, 13, 25]),
 (66, [9, 21, 26, 56]),
 (68, [0]),
 (69, [0, 9, 25, 26, 39, 67]),
 (70, [0]),
 (71, [0]),
 (72, [3, 9, 25, 26, 40, 56, 67]),
 (73, [0, 1, 25, 30, 48]),
 (74, [0]),
 (75, [9, 56, 67]),
 (76, [0, 25]),
 (77, [0, 40]),
 (78, [0, 4]),
 (79, [9, 25, 26, 67]),
 (80, [57]),
 (81, [0]),
 (82, [13, 16, 29, 56, 59, 65, 67]),
 (83, [0, 23, 25, 31, 51]),


In [36]:
# cut_sort_friends_list's 0 doesn't have 1
len(number_of_friends.filter(lambda x : x[1] < 17 and x[0] <= 347).collect())

235

## Broadcast friends

마지막 Jaccard 연산 수행을위해 실제 전체 친구 리스트가 필요하므로, way2_friends를 broadcast

In [37]:
broad_friends_list = sc.broadcast(way2_friends.collect())

In [38]:
broad_friends_list.value[:10]

[(0,
  [1,
   2,
   3,
   4,
   5,
   6,
   7,
   8,
   9,
   10,
   11,
   12,
   13,
   14,
   15,
   16,
   17,
   18,
   19,
   20,
   21,
   22,
   23,
   24,
   25,
   26,
   27,
   28,
   29,
   30,
   31,
   32,
   33,
   34,
   35,
   36,
   37,
   38,
   39,
   40,
   41,
   42,
   43,
   44,
   45,
   46,
   47,
   48,
   49,
   50,
   51,
   52,
   53,
   54,
   55,
   56,
   57,
   58,
   59,
   60,
   61,
   62,
   63,
   64,
   65,
   66,
   67,
   68,
   69,
   70,
   71,
   72,
   73,
   74,
   75,
   76,
   77,
   78,
   79,
   80,
   81,
   82,
   83,
   84,
   85,
   86,
   87,
   88,
   89,
   90,
   91,
   92,
   93,
   94,
   95,
   96,
   97,
   98,
   99,
   100,
   101,
   102,
   103,
   104,
   105,
   106,
   107,
   108,
   109,
   110,
   111,
   112,
   113,
   114,
   115,
   116,
   117,
   118,
   119,
   120,
   121,
   122,
   123,
   124,
   125,
   126,
   127,
   128,
   129,
   130,
   131,
   132,
   133,
   134,
   135,
   136,
   137,
   138,

## Jaccard Similarity

cal_jaccard : python의 set 연산을 통해 연산 수행 <br>
find_similar : jaccard 연산을 수행하는 함수
> broad_friends_list는 key값이 오름차순으로 정렬되어 있으므로 index를 통해 불러올 수 있음 <br>
> key1 < key2 일 때만 비교함으로써 중복연산을 제거할 수 있음 <br>
> 중복되는 추천 친구가 발생할 수 있으므로 ((key1, key2), Jaccard_coef)형태로 key pair를 분리

tmp_result : jaccard값을 단순 계산한 임시 결과 <br>
result : threshold를 넘는 jaccard값을 필터링하여 중복을 제거하면 최종 결과가 됨 <br>

In [39]:
def cal_jaccard(vec1, vec2) : 
    set1 = set(vec1)
    set2 = set(vec2)
    return len(set1.intersection(set2)) / len(set1.union(set2))

In [40]:
print(broad_friends_list.value[4])
print(broad_friends_list.value[5])
print(cal_jaccard(broad_friends_list.value[4][1], broad_friends_list.value[5][1]))

(4, [0, 78, 152, 181, 195, 218, 273, 275, 306, 328])
(5, [0, 87, 122, 156, 158, 169, 180, 187, 204, 213, 235, 315, 316])
0.045454545454545456


In [41]:
def find_similar(input_data) : 
    friends_list = [broad_friends_list.value[num] for num in input_data[1]]
    tmp_list = []
    for vec1 in friends_list : 
        for vec2 in friends_list : 
            if (vec1[0] < vec2[0]) : 
                jaccard_coef = cal_jaccard(vec1[1], vec2[1])
                tmp_list.append(((vec1[0], vec2[0]), jaccard_coef))
    return tmp_list

In [42]:
tmp_result = inverted_friends_list.flatMap(lambda x : find_similar(x))

In [43]:
tmp_result = tmp_result.sortBy(lambda x : x[0])

In [44]:
result = tmp_result.filter(lambda x : x[1] > CFG.sigma)

In [45]:
result.collect()

[((2, 14), 0.5625),
 ((2, 14), 0.5625),
 ((2, 14), 0.5625),
 ((2, 14), 0.5625),
 ((2, 149), 0.6),
 ((2, 149), 0.6),
 ((2, 162), 0.8),
 ((2, 226), 0.6),
 ((4, 78), 0.7272727272727273),
 ((4, 78), 0.7272727272727273),
 ((4, 78), 0.7272727272727273),
 ((4, 181), 0.8181818181818182),
 ((4, 181), 0.8181818181818182),
 ((4, 181), 0.8181818181818182),
 ((4, 195), 0.7272727272727273),
 ((4, 195), 0.7272727272727273),
 ((4, 218), 0.7272727272727273),
 ((6, 89), 0.5555555555555556),
 ((6, 89), 0.5555555555555556),
 ((6, 147), 0.7142857142857143),
 ((7, 339), 0.5161290322580645),
 ((7, 339), 0.5161290322580645),
 ((8, 91), 0.7777777777777778),
 ((8, 91), 0.7777777777777778),
 ((8, 91), 0.7777777777777778),
 ((8, 91), 0.7777777777777778),
 ((9, 26), 0.5432098765432098),
 ((9, 26), 0.5432098765432098),
 ((9, 26), 0.5432098765432098),
 ((9, 26), 0.5432098765432098),
 ((9, 26), 0.5432098765432098),
 ((9, 26), 0.5432098765432098),
 ((9, 26), 0.5432098765432098),
 ((9, 26), 0.5432098765432098),
 ((9, 2

In [46]:
result.distinct().collect()

[((2, 14), 0.5625),
 ((2, 149), 0.6),
 ((2, 162), 0.8),
 ((2, 226), 0.6),
 ((4, 78), 0.7272727272727273),
 ((4, 181), 0.8181818181818182),
 ((4, 195), 0.7272727272727273),
 ((4, 218), 0.7272727272727273),
 ((6, 89), 0.5555555555555556),
 ((6, 147), 0.7142857142857143),
 ((7, 339), 0.5161290322580645),
 ((8, 91), 0.7777777777777778),
 ((9, 26), 0.5432098765432098),
 ((9, 188), 0.5909090909090909),
 ((13, 98), 0.5094339622641509),
 ((13, 118), 0.5227272727272727),
 ((13, 325), 0.5555555555555556),
 ((14, 28), 0.6470588235294118),
 ((14, 116), 0.5238095238095238),
 ((14, 140), 0.5294117647058824),
 ((14, 144), 0.5789473684210527),
 ((14, 149), 0.5263157894736842),
 ((17, 28), 0.5294117647058824),
 ((17, 140), 0.6),
 ((20, 116), 0.6),
 ((20, 149), 0.5263157894736842),
 ((20, 226), 0.5263157894736842),
 ((21, 67), 0.5161290322580645),
 ((21, 122), 0.5609756097560976),
 ((21, 252), 0.5116279069767442),
 ((24, 92), 0.6818181818181818),
 ((24, 94), 0.5833333333333334),
 ((24, 101), 0.666666666

## 친구 list의 이해

A B C D E <br>
1 2 3 4 5 <br>

#### 단방향 (페북식)
1 - 3 4 5 <br>
2 - 4 5 <br>
3 - 5 <br>
4 -  <br>
5 -  <br>

#### 양방향 (변경)
1 - 3 4 5 <br>
2 - 4 5 <br>
3 - 1 5 <br>
4 - 1 2 <br>
5 - 1 2 3 <br>

#### 양방향 inverted list
1 - 3 4 5 <br>
2 - 4 5 <br>
3 - 1 5 <br>
4 - 1 2 <br>
5 - 1 2 3 <br>

=> 양방향의 inverted list는 동일함 <br>

#### 단방향 inverted list
1 -  <br>
2 -  <br>
3 - 1  <br>
4 - 1 2 <br>
5 - 1 2 3  <br>

따라서 기존 facebook_combined 데이터를 이용하면 중복 연산을 줄일 수 있을 것으로 예상 <br>
> 단, freq vocab은 전체 빈도를 기준으로 하므로 양방향 데이터를 생성해야함

## 중복 연산 체크

find_similar 함수의 tmp_list.append(((vec1[0], vec2[0]), jaccard_coef))를 수정하여 <br>
어떤 item의 element 조사 중 찾아진 pair인지 추출하면 전부 다름 <br> 
<br>
((2, 14), 0.5625, 20), <br>
((2, 14), 0.5625, 116), <br>
((2, 14), 0.5625, 149), <br>
((2, 14), 0.5625, 226), <br>
((2, 149), 0.6, 226), <br>
((2, 149), 0.6, 333), <br>
((2, 162), 0.8, 333), <br>
((2, 226), 0.6, 333) <br>